<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>ComutR • ComutR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="ComutR">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">ComutR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="active nav-item"><a class="nav-link" href="../articles/ComutR.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>ComutR</h1>
            
      

      <div class="d-none name"><code>ComutR.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://taprati.github.io/ComutR/">ComutR</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="preparing-data-for-comut-plots">Preparing data for comut plots<a class="anchor" aria-label="anchor" href="#preparing-data-for-comut-plots"></a>
</h2>
<p>The main input to comut is a dataframe with columns found in a
typical maf file. Three columns are required by comut:</p>
<ul>
<li>Tumor_Sample_Barcode</li>
<li>Hugo_Symbol</li>
<li>Variant_Classification</li>
</ul>
<p>For more information on the maf format and column names, visit <a href="https://docs.gdc.cancer.gov/Data/File_Formats/MAF_Format/" class="external-link">maf
format</a></p>
<div class="section level3">
<h3 id="generating-mock-data">Generating mock data<a class="anchor" aria-label="anchor" href="#generating-mock-data"></a>
</h3>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">1999</span><span class="op">)</span></span>
<span></span>
<span><span class="va">sample_ids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">)</span></span>
<span></span>
<span><span class="va">gene_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"B"</span>, <span class="st">"C"</span>, <span class="st">"D"</span>, <span class="st">"E"</span>, <span class="st">"F"</span>, <span class="st">"G"</span>, <span class="st">"H"</span><span class="op">)</span></span>
<span><span class="va">gene_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"gene_"</span>, <span class="va">gene_names</span><span class="op">)</span></span>
<span></span>
<span><span class="va">variant_types</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Frame_Shift_Del"</span>, <span class="st">"Frame_Shift_Ins"</span>, <span class="st">"In_Frame_Del"</span>, <span class="st">"In_Frame_Ins"</span>, <span class="st">"Missense_Mutation"</span>, <span class="st">"Nonsense_Mutation"</span>, <span class="st">"Splice_Site"</span>, <span class="st">"Translation_Start_Site"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">maf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/expand.grid.html" class="external-link">expand.grid</a></span><span class="op">(</span></span>
<span>  <span class="st">"Tumor_Sample_Barcode"</span> <span class="op">=</span> <span class="va">sample_ids</span>,</span>
<span>  <span class="st">"Hugo_Symbol"</span> <span class="op">=</span> <span class="va">gene_names</span></span>
<span>  <span class="op">)</span></span>
<span><span class="va">maf</span><span class="op">$</span><span class="va">Variant_Classification</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="va">variant_types</span>,</span>
<span>                                     size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">maf</span><span class="op">)</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Randomly subset</span></span>
<span><span class="va">input_maf</span> <span class="op">&lt;-</span> <span class="va">maf</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span></span>
<span>  <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/sample_n.html" class="external-link">sample_frac</a></span><span class="op">(</span><span class="fl">0.35</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Duplicate first row twice for showing multiple</span></span>
<span><span class="va">input_maf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">input_maf</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span>,<span class="op">]</span>, <span class="va">input_maf</span><span class="op">)</span></span>
<span><span class="co"># Duplicate last row once for showing double annotations</span></span>
<span><span class="va">input_maf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">input_maf</span><span class="op">[</span><span class="fl">4</span>,<span class="op">]</span>, <span class="va">input_maf</span><span class="op">)</span></span>
<span><span class="va">input_maf</span><span class="op">[</span><span class="fl">1</span>, <span class="fl">3</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"Missense_Mutation"</span></span>
<span></span>
<span><span class="va">input_maf</span> <span class="op"><a href="../reference/pipe.html">%&gt;%</a></span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##     Tumor_Sample_Barcode Hugo_Symbol Variant_Classification</span></span>
<span><span class="co">## 2                      8      gene_A      Missense_Mutation</span></span>
<span><span class="co">## 1                      7      gene_H           In_Frame_Ins</span></span>
<span><span class="co">## 1.1                    7      gene_H           In_Frame_Ins</span></span>
<span><span class="co">## 110                    7      gene_H           In_Frame_Ins</span></span>
<span><span class="co">## 29                     8      gene_A            Splice_Site</span></span>
<span><span class="co">## 3                      3      gene_A           In_Frame_Ins</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="basic-comut-plots">Basic Comut plots<a class="anchor" aria-label="anchor" href="#basic-comut-plots"></a>
</h2>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/comut.html">comut</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">input_maf</span><span class="op">)</span></span></code></pre></div>
<p><img src="ComutR_files/figure-html/basic%20comut-1.png" width="480"></p>
<p>By default, comutR provides a colorscheme for the variant
classification, but this can be altered using the
<code>variant_colors</code> and <code>variant_scheme</code> arguments.
<code>comut</code> also orders the rows by frequency of alterations.</p>
</div>
<div class="section level2">
<h2 id="adding-metadata">Adding metadata<a class="anchor" aria-label="anchor" href="#adding-metadata"></a>
</h2>
<p>Metadata can be added through the metadata argument of comut. This
expects and dataframe with Tumor_Sample_Barcode and then additional
columns for the metadata.</p>
<p>Along with metadata, you must pass the col_maps argument, which
contains a named list of color maps corresponding to the columns in the
metadata.</p>
<p>Here we have gender and biopsy site information for each sample, and
we define color maps for the variables accordingly:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">meta</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="st">"Tumor_Sample_Barcode"</span> <span class="op">=</span> <span class="va">sample_ids</span><span class="op">)</span></span>
<span><span class="va">meta</span><span class="op">$</span><span class="va">biopsy</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Primary"</span>, <span class="st">"Metastasis"</span><span class="op">)</span>, size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">meta</span><span class="op">)</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">meta</span><span class="op">$</span><span class="va">sex</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Male"</span>, <span class="st">"Female"</span><span class="op">)</span>, size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">meta</span><span class="op">)</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Define color mapping</span></span>
<span><span class="va">sex_colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"Male"</span> <span class="op">=</span> <span class="st">"lightblue"</span>,</span>
<span>  <span class="st">"Female"</span> <span class="op">=</span> <span class="st">"pink"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">biopsy_type_colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"Primary"</span> <span class="op">=</span> <span class="st">"lightgreen"</span>,</span>
<span>  <span class="st">"Metastasis"</span> <span class="op">=</span> <span class="st">"maroon"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">color_maps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  <span class="st">"sex"</span> <span class="op">=</span> <span class="va">sex_colors</span>,</span>
<span>  <span class="st">"biopsy"</span> <span class="op">=</span> <span class="va">biopsy_type_colors</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Note that patients without metadata will be colored grey.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/comut.html">comut</a></span><span class="op">(</span>data <span class="op">=</span> <span class="va">input_maf</span>, metadata <span class="op">=</span> <span class="va">meta</span>, col_maps <span class="op">=</span> <span class="va">color_maps</span><span class="op">)</span></span></code></pre></div>
<p><img src="ComutR_files/figure-html/Adding%20metadata-1.png" width="480"></p>
</div>
<div class="section level2">
<h2 id="adding-barplots">Adding Barplots<a class="anchor" aria-label="anchor" href="#adding-barplots"></a>
</h2>
<p>Barplot data is formatted in a similar manner to metadata. A column
for <code>Tumor_Sample_Barcode</code> is required, and then columns for
the barplot values. If a stacked barplot is desired (for mutation
signatures for example) then there should be multiple columns
corresponding to the proportion of each value.</p>
<p>Barplots are defined within a named list of named lists. The top
level name of each sublist is used as the y-axis label for the barplot.
Each sublist should contain “data” containing the barplot data, “colors”
containing a color map for the barplot, and “legend” a boolean for
whether a legend should be created.</p>
<p>Note: The order of the barplot data list is the reverse of the order
in which they are displayed in the plot!</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">signatures</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="st">"Tumor_Sample_Barcode"</span> <span class="op">=</span> <span class="va">sample_ids</span><span class="op">)</span></span>
<span><span class="va">signatures</span><span class="op">$</span><span class="va">Aging</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">meta</span><span class="op">)</span>, min <span class="op">=</span> <span class="fl">0</span>, max <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">signatures</span><span class="op">$</span><span class="va">Other</span> <span class="op">&lt;-</span> <span class="fl">1</span> <span class="op">-</span> <span class="va">signatures</span><span class="op">$</span><span class="va">Aging</span></span>
<span></span>
<span><span class="va">total_snp_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="st">"Tumor_Sample_Barcode"</span> <span class="op">=</span> <span class="va">sample_ids</span><span class="op">)</span></span>
<span><span class="va">total_snp_data</span><span class="op">$</span><span class="va">n_snps</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">20</span>, size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">meta</span><span class="op">)</span>, replace <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span></span>
<span><span class="va">signature_colors</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="st">"Aging"</span> <span class="op">=</span> <span class="st">"coral1"</span>,</span>
<span>  <span class="st">"Other"</span> <span class="op">=</span> <span class="st">"antiquewhite"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Barplot specifications</span></span>
<span><span class="va">barplot_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>  <span class="st">"Mutation Signatures"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    <span class="st">"data"</span> <span class="op">=</span> <span class="va">signatures</span>,</span>
<span>    <span class="st">"colors"</span> <span class="op">=</span> <span class="va">signature_colors</span>,</span>
<span>    <span class="st">"legend"</span> <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>    <span class="op">)</span>,</span>
<span>  <span class="st">"N SNPs"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    <span class="st">"data"</span> <span class="op">=</span> <span class="va">total_snp_data</span>,</span>
<span>    <span class="st">"colors"</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"n_snps"</span> <span class="op">=</span> <span class="st">"cadetblue3"</span><span class="op">)</span>,</span>
<span>    <span class="st">"legend"</span> <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/comut.html">comut</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="va">input_maf</span>,</span>
<span>  barplot_data <span class="op">=</span> <span class="va">barplot_data</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="ComutR_files/figure-html/unnamed-chunk-3-1.png" width="480"></p>
<p>Putting all of the components together, we get something like
this:</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/comut.html">comut</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="va">input_maf</span>,</span>
<span>  metadata <span class="op">=</span> <span class="va">meta</span>,</span>
<span>  col_maps <span class="op">=</span> <span class="va">color_maps</span>,</span>
<span>  barplot_data <span class="op">=</span> <span class="va">barplot_data</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="ComutR_files/figure-html/unnamed-chunk-4-1.png" width="480"></p>
</div>
<div class="section level2">
<h2 id="text-annotations">Text Annotations<a class="anchor" aria-label="anchor" href="#text-annotations"></a>
</h2>
<p>ComutR allows the user to add arbitrary text annotations to the cells
of the comut plot. This can be specified as a column name that appears
in the input data. Here we show adding allele fraction annotations to
each mutation:</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">input_maf</span><span class="op">$</span><span class="va">Allele_Fraction</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Round.html" class="external-link">round</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/Uniform.html" class="external-link">runif</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">input_maf</span><span class="op">)</span>, min <span class="op">=</span> <span class="fl">0.1</span>, max <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>, <span class="fl">1</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="../reference/comut.html">comut</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="va">input_maf</span>,</span>
<span>  text_annotation <span class="op">=</span> <span class="st">"Allele_Fraction"</span></span>
<span>  <span class="op">)</span></span></code></pre></div>
<p><img src="ComutR_files/figure-html/unnamed-chunk-5-1.png" width="480"></p>
</div>
<div class="section level2">
<h2 id="limiting-samples-or-features-and-filling-missing-information">Limiting samples or features and filling missing information<a class="anchor" aria-label="anchor" href="#limiting-samples-or-features-and-filling-missing-information"></a>
</h2>
<p>If a subset of samples or features in the input data are desired, the
user can specify the <code>ids</code> or
<code>features_of_interest</code> arguments respectively.
<code>comut</code> will also fill in any ids or features that are not
present in the input data. This can be especially useful when one wants
to highlight a feature that is not altered in any samples.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/comut.html">comut</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="va">input_maf</span>,</span>
<span>  ids <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"1"</span>, <span class="st">"3"</span>, <span class="st">"4"</span>, <span class="st">"7"</span>, <span class="st">"missing_sample"</span><span class="op">)</span>,</span>
<span>  features_of_interest <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gene_A"</span>, <span class="st">"gene_B"</span>, <span class="st">"gene_E"</span>, <span class="st">"gene_H"</span>, <span class="st">"gene_missing"</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="ComutR_files/figure-html/unnamed-chunk-6-1.png" width="480"></p>
</div>
<div class="section level2">
<h2 id="color-schemes">Color schemes<a class="anchor" aria-label="anchor" href="#color-schemes"></a>
</h2>
<p>By default, <code>comut</code> assigns colors from the <a href="https://loading.io/color/feature/Vivid-11" class="external-link">vivid</a> color scheme
to the variant classifications in order of frequency. To specify a
custom color scheme, the <code>variant_colors</code> argument is used.
This can be used to explicitly show variant not seen in the dataset.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/comut.html">comut</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="va">input_maf</span>,</span>
<span>  ids <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"1"</span>, <span class="st">"3"</span>, <span class="st">"4"</span>, <span class="st">"7"</span>, <span class="st">"missing_sample"</span><span class="op">)</span>,</span>
<span>  features_of_interest <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gene_A"</span>, <span class="st">"gene_B"</span>, <span class="st">"gene_E"</span>, <span class="st">"gene_H"</span>, <span class="st">"gene_missing"</span><span class="op">)</span>,</span>
<span>  variant_colors <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>    <span class="st">"In_Frame_Ins"</span> <span class="op">=</span> <span class="st">"lightblue"</span>,</span>
<span>    <span class="st">"Frame_Shift_Ins"</span> <span class="op">=</span> <span class="st">"pink"</span>,</span>
<span>    <span class="st">"Unseen Variant Class"</span> <span class="op">=</span> <span class="st">"purple"</span>,</span>
<span>    <span class="st">"In_Frame_Del"</span> <span class="op">=</span> <span class="st">"lightgreen"</span></span>
<span>  <span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="ComutR_files/figure-html/unnamed-chunk-7-1.png" width="480"></p>
</div>
<div class="section level2">
<h2 id="additional-options">Additional Options<a class="anchor" aria-label="anchor" href="#additional-options"></a>
</h2>
<p>Comut also provide many arguments for the plot size, fontsize, and
other options.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/comut.html">comut</a></span><span class="op">(</span></span>
<span>  data <span class="op">=</span> <span class="va">input_maf</span>,</span>
<span>  metadata <span class="op">=</span> <span class="va">meta</span>,</span>
<span>  col_maps <span class="op">=</span> <span class="va">color_maps</span>,</span>
<span>  barplot_data <span class="op">=</span> <span class="va">barplot_data</span>,</span>
<span>  text_annotation <span class="op">=</span> <span class="st">"Allele_Fraction"</span>,</span>
<span>  cell_width <span class="op">=</span> <span class="fl">0.5</span>,</span>
<span>  cell_height <span class="op">=</span> <span class="fl">0.3</span>,</span>
<span>  anno_fontsize <span class="op">=</span> <span class="fl">12</span>,</span>
<span>  legend_fontsize <span class="op">=</span> <span class="fl">12</span>,</span>
<span>  add_borders <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  body_border <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>  <span class="co"># row_names_rot = 30,</span></span>
<span>  <span class="co"># column_names_rot = 0</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p><img src="ComutR_files/figure-html/unnamed-chunk-8-1.png" width="960"></p>
<p><code>comut</code> also takes additional variable arguments for the
call to <code>Heatmap</code>, so users can specify existing options like
<code>row_names_rot</code>, <code>column_names_rot</code>,
<code>body_border</code> etc.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Tyler Aprati.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
